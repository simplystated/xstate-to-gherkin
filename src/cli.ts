import * as meta from "@xstate/machine-extractor/src/"
import * as arg from "arg";
import { existsSync, readFileSync, writeFileSync } from "fs";
import { createMachine } from "xstate";
import { parseMachinesFromFile } from "@xstate/machine-extractor";
import { toGherkinScripts, xstateToGherkin } from ".";
import * as path from "node:path";

export const run = (argv: Array<string>): number => {
  const args = arg(
    {
      "--help": Boolean,
      "--version": Boolean,
      "--output-dir": String,
      "--machine": String,
      "--force": Boolean,

      // aliases
      "-h": "--help",
      "-v": "--version",
      "-o": "--output-dir",
      "-m": "--machine",
      "-f": "--force",
    },
    {
        stopAtPositional: true,
        argv
    }
  );

  if ( args["--help"] ) {
    usage();
    return 0;
  }

  if ( args["--version"] ) {
    const packagejson = JSON.parse(readFileSync(path.join(__dirname, "..", "package.json"), "utf8"));
    console.log(`
        xstate-to-gherkin v${packagejson.version}
    `)
    return 0;
  }

  const outputDir = args["--output-dir"];
  const positionalArgs = args["_"]
  const selectedMachine = args["--machine"];
  const force = !!args["--force"];

  if ( !outputDir ) {
    console.error("--output-dir is required");
    usage();
    return 1;
  }

  if ( !positionalArgs || positionalArgs.length !== 1 ) {
    console.error("path to js/ts file is required (pass it after all other arguments)");
    usage();
    return 1;
  }

  return main(outputDir, positionalArgs[0], selectedMachine ?? null, force);
};

const usage = () => {
    console.log(`
        xstate-to-gherkin converts annotated xstate statecharts into gherkin test scripts.

        Usage:
          xstate-to-gherkin [--machine-name <name-of-machine-in-js-file>] [--force] --output-dir <directory-for-generated-gherkin-scripts> <path-to-js-file-containing-annotated-xstate-state-machine>

          Generates a set of .feature files in <directory-for-generated-gherkin-scripts> by traversing all simple paths in the statechart
          contained in <path-to-js-file-containing-annotated-xstate-state-machine>. If multiple machines are present in 
          <path-to-js-file-containing-annotated-xstate-state-machine>, specify --machine-name to pick one to use.
          By default, xstate-to-gherkin will not overwrite any files. Pass --force to overwrite.

        Statechart annotation:
          This tool requires each state to be annotated with up to 3 meta fields:
            gherkinFeature - (optional) specifies the feature this state is a part of
            gherkinScenario - (optional) specifies the scenario this state is a part of
            gherkinAssert - specifies the assertion that should hold in this state
    `);
}

const main = (outputDir: string, inputPath: string, selectedMachine: string | null, force: boolean): number => {
    const js = readFileSync(inputPath, "utf8");
    const machines = parseMachinesFromFile(js);
    const machineConfigs = machines.machines.map(m => m.toConfig());
    if ( machineConfigs.length > 1 && !selectedMachine) {
        console.error(`multiple machines found at ${inputPath}. specify --machine to select one.`);
        return 1;
    }
    const machineConfig = selectedMachine ? machineConfigs.filter(c => c?.id === selectedMachine)[0] : machineConfigs[0];
    if ( !machineConfig ) {
        if ( selectedMachine ) {
            console.error(`no machine named ${selectedMachine} found in ${inputPath}. found machines: [${machineConfigs.map(m => m?.id).join(", ")}].`)
        } else {
            console.error(`no machines found in ${inputPath}.`);
        }
        return 1;
    }

    const scriptsByName = toGherkinScripts(xstateToGherkin(createMachine({ ...machineConfig, predictableActionArguments: true })));
    for (const [filename, script] of scriptsByName.entries()) {
        const filePath = path.join(outputDir, filename);
        if ( !force && existsSync(filePath) ) {
            console.error(`refusing to overwrite existing file ${filePath}. pass --force to overwrite.`);
            return 1;
        }
        writeFileSync(path.join(outputDir, filename), script, { encoding: "utf8" });
    }

    return 0;
};
