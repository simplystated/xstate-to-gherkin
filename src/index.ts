import { getSimplePaths } from "@xstate/graph";
import {
  AnyEventObject,
  createMachine,
  Segment,
  State,
  StateMachine,
  StateNode,
  StateNodeDefinition,
} from "xstate";

type AnyMachine = StateMachine<any, any, any>;
type AnyState = StateNodeDefinition<any, any, any>;

export const xstateToGherkin = (
  machine: AnyMachine
): Array<Array<GherkinStep>> => {
  const canonicalDefn = machine.definition;
  const flattened = flattenConds(canonicalDefn);
  const pathsByEndState = getSimplePaths(
    createMachine({ predictableActionArguments: true, ...flattened })
  );
  return Object.keys(pathsByEndState)
    .flatMap((endState) => {
      const { state, paths } = pathsByEndState[endState];
      return paths.map((path) =>
        path.segments.flatMap((segment, idx) => {
          const gherkinStepsForState = gherkinStepsFromState(
            segment.state,
            idx === 0
          );
          const gherkinStepsForEvent = gherkinStepsFromEvent(segment.event);
          return gherkinStepsForState
            .concat(gherkinStepsForEvent)
            .concat(
              gherkinStepsFromState(path.state, path.segments.length === 0)
            );
        })
      );
    })
    .filter((steps) => steps.length > 0);
};

export interface GherkinStep {
  keyword: "Given" | "Then" | "And" | "When";
  step: string;
}

const gherkinStepsFromEvent = (event: AnyEventObject): Array<GherkinStep> => [
  { keyword: "When", step: event.type.replace(/_/g, " ") },
];

const gherkinStepsFromState = (
  state: State<any, any>,
  isFirstStep: boolean
): Array<GherkinStep> =>
  valuesSortedByKey(state.meta)
    .map((m) => (m as any).gherkinAssert)
    .filter((gherkinAssert) => !!gherkinAssert)
    .map((step, idx) => {
      const keyword =
        idx === 0 && isFirstStep ? "Given" : idx === 0 ? "Then" : "And";
      return { keyword, step };
    });

const valuesSortedByKey = <T>(m: Record<string, T>): Array<T> => {
  const keys = Object.keys(m);
  keys.sort();
  return keys.map((k) => m[k]);
};

const flattenConds = (state: AnyState): AnyState => {
  const { transitions, ...stateDefinition } = state;

  stateDefinition.on = Object.keys(stateDefinition.on).reduce((on, evtName) => {
    const evts = stateDefinition.on[evtName];
    return evts.reduce((on, evt) => {
      const { cond, ...unconditionalEvt } = evt;
      if (cond) {
        const condName = cond.name;
        const newEvtName = `${evtName} with ${condName}`;
        return {
          ...on,
          [newEvtName]: unconditionalEvt,
        };
      }

      return {
        ...on,
        [evtName]: evt,
      };
    }, on);
  }, {});

  stateDefinition.states = Object.keys(stateDefinition.states).reduce(
    (states, stateName) => ({
      ...states,
      [stateName]: flattenConds(stateDefinition.states[stateName]),
    }),
    {}
  );

  return new StateNode(stateDefinition).definition;
};
