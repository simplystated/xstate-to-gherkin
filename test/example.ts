import { createMachine } from "xstate";
import {
  AstBuilder,
  GherkinClassicTokenMatcher,
  Parser,
} from "@cucumber/gherkin";
import { IdGenerator } from "@cucumber/messages";
import { toGherkinScripts, xstateToGherkin } from "../src";

const basicWithConditions = createMachine({
  predictableActionArguments: true,
  id: "basicWithConditions",
  initial: "start",
  states: {
    start: {
      meta: {
        gherkinAssert: "I can see the start page",
        gherkinFeature: "My Feature",
      },
      on: {
        "I tap the next button": [
          {
            target: "success",
            cond: "good input",
          },
          {
            target: "failure",
            cond: "bad input",
          },
        ],
      },
    },
    success: {
      meta: {
        gherkinAssert: "I can see the success page",
        gherkinScenario: "Success",
      },
      initial: "successStep1",
      states: {
        successStep1: {
          meta: {
            gherkinAssert: "I see the first step of the success page",
          },
        },
      },
    },
    failure: {
      meta: {
        gherkinAssert: "I see the failure page",
        gherkinScenario: "Failure",
      },
    },
  },
});

const basicWithTransientTransitions = createMachine({
  predictableActionArguments: true,
  id: "basicWithTransientTransitions",
  initial: "start",
  states: {
    start: {
      meta: {
        gherkinAssert: "I can see the start page",
        gherkinFeature: "My Feature",
      },
      always: [
        { target: "success", cond: "good input" },
        { target: "failure", cond: "bad input" },
      ],
    },
    success: {
      meta: {
        gherkinAssert: "I can see the success page",
        gherkinScenario: "Success",
      },
      initial: "successStep1",
      states: {
        successStep1: {
          meta: {
            gherkinAssert: "I see the first step of the success page",
          },
        },
      },
    },
    failure: {
      meta: {
        gherkinAssert: "I see the failure page",
        gherkinScenario: "Failure",
      },
    },
  },
});

describe("xstate to gherkin", () => {
  it("works for a basic example with conditions", () => {
    expect(xstateToGherkin(basicWithConditions)).toMatchSnapshot();
  });

  it("works for a basic example with transient transitions", () => {
    expect(xstateToGherkin(basicWithTransientTransitions)).toMatchSnapshot();
  });
});

describe("xstate to gherkin script", () => {
  it("works for a basic example with conditions", () => {
    const scripts = toGherkinScripts(xstateToGherkin(basicWithConditions));

    const parser = gherkinParser();
    scripts.forEach((script) => {
      expect(parser.parse(script).feature?.name).toEqual("My Feature");
    });

    expect(scripts).toMatchSnapshot();
  });
});

const gherkinParser = () => {
  const uuidFn = IdGenerator.uuid();
  const builder = new AstBuilder(uuidFn);
  const matcher = new GherkinClassicTokenMatcher();

  return new Parser(builder, matcher);
};
