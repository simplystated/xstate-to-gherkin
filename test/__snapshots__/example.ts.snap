// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xstate to gherkin script works for a basic example with conditions 1`] = `
Map {
  "my-feature.feature" => "Feature: My Feature

  Scenario: Success
    Given I can see the start page
    When I tap the next button with good input
    Then I can see the success page
    And I see the first step of the success page

  Scenario: Failure
    Given I can see the start page
    When I tap the next button with bad input
    Then I see the failure page",
}
`;

exports[`xstate to gherkin works for a basic example with conditions 1`] = `
[
  {
    "feature": "My Feature",
    "scenarios": [
      {
        "scenario": "Success",
        "steps": [
          {
            "keyword": "Given",
            "step": "I can see the start page",
          },
          {
            "keyword": "When",
            "step": "I tap the next button with good input",
          },
          {
            "keyword": "Then",
            "step": "I can see the success page",
          },
          {
            "keyword": "And",
            "step": "I see the first step of the success page",
          },
        ],
      },
      {
        "scenario": "Failure",
        "steps": [
          {
            "keyword": "Given",
            "step": "I can see the start page",
          },
          {
            "keyword": "When",
            "step": "I tap the next button with bad input",
          },
          {
            "keyword": "Then",
            "step": "I see the failure page",
          },
        ],
      },
    ],
  },
]
`;

exports[`xstate to gherkin works for a basic example with transient transitions 1`] = `
[
  {
    "feature": "My Feature",
    "scenarios": [
      {
        "scenario": "Success",
        "steps": [
          {
            "keyword": "Given",
            "step": "I can see the start page",
          },
          {
            "keyword": "When",
            "step": "good input",
          },
          {
            "keyword": "Then",
            "step": "I can see the success page",
          },
          {
            "keyword": "And",
            "step": "I see the first step of the success page",
          },
        ],
      },
      {
        "scenario": "Failure",
        "steps": [
          {
            "keyword": "Given",
            "step": "I can see the start page",
          },
          {
            "keyword": "When",
            "step": "bad input",
          },
          {
            "keyword": "Then",
            "step": "I see the failure page",
          },
        ],
      },
    ],
  },
]
`;
